File: spark_example_kafka_to_parquet.py
from pyspark.sql import SparkSession
from pyspark.sql.functions import from_json, col
from pyspark.sql.types import StructType, StructField, IntegerType, StringType, DoubleType

spark = SparkSession.builder \
    .appName("KafkaToParquet") \
    .getOrCreate()

# Define schema for the JSON payload
schema = StructType([
    StructField("id", IntegerType()),
    StructField("name", StringType()),
    StructField("amount", DoubleType()),
    StructField("ts", StringType())
])

# Read from Kafka
df_kafka = spark.read \
    .format("kafka") \
    .option("kafka.bootstrap.servers", "kafka:9092") \
    .option("subscribe", "kitchensink") \
    .option("startingOffsets", "earliest") \
    .load()

# Parse JSON and select fields
df = df_kafka.select(from_json(col("value").cast("string"), schema).alias("data")).select("data.*")

# Write as Parquet to MinIO (S3)
df.write.mode("overwrite").parquet("s3a://warehouse/kitchensink")

spark.stop()

File: spark_example_kafka_to_parquet.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.types._
import org.apache.spark.sql.functions._

object KafkaToParquet {
  def main(args: Array[String]): Unit = {
    val spark = SparkSession.builder
      .appName("KafkaToParquet")
      .getOrCreate()

    import spark.implicits._

    // Define the schema for the JSON payload
    val schema = StructType(Seq(
      StructField("id", IntegerType, true),
      StructField("name", StringType, true),
      StructField("amount", DoubleType, true),
      StructField("ts", StringType, true)
    ))

    // Read from Kafka
    val kafkaDF = spark.read
      .format("kafka")
      .option("kafka.bootstrap.servers", "kafka:9092")
      .option("subscribe", "kitchensink")
      .option("startingOffsets", "earliest")
      .load()

    // Parse the JSON and select fields
    val jsonDF = kafkaDF.select(
      from_json(col("value").cast("string"), schema).alias("data")
    ).select("data.*")

    // Write as Parquet to MinIO (S3)
    jsonDF.write.mode("overwrite").parquet("s3a://warehouse/kitchensink")

    spark.stop()
  }
}